lib/yargs.ts

```diff
@@ -0,0 +1,1694 @@
+            self.exit(0)
+          }
+        })
+      }
+
+      // Check if any of the options to skip validation were provided
+      if (!skipValidation && options.skipValidation.length > 0) {
+        skipValidation = Object.keys(argv).some(key => options.skipValidation.indexOf(key) >= 0 && argv[key] === true)
+      }
+
+      // If the help or version options where used and exitProcess is false,
+      // or if explicitly skipped, we won't run validations.
+      if (!skipValidation) {
+        if (parsed.error) throw new YError(parsed.error.message)
+
+        // if we're executed via bash completion, don't
+        // bother with validation.
+        if (!requestCompletions) {
+          self._runValidation(argv, aliases, {}, parsed.error)
+        }
+      }
+    } catch (err) {
+      if (err instanceof YError) usage.fail(err.message, err)
+      else throw err
+    }
+
+    return (populateDoubleDash || _calledFromCommand) ? argv : self._copyDoubleDash(argv)
+  }
+
+  // to simplify the parsing of positionals in commands,
+  // we temporarily populate '--' rather than _, with arguments
+  // after the '--' directive. After the parse, we copy these back.
+  self._copyDoubleDash = function (argv: Arguments | Promise<Arguments>): any {
+    if (isPromise(argv) || !argv._ || !argv['--']) return argv
+    argv._.push.apply(argv._, argv['--'])
+
+    // TODO(bcoe): refactor command parsing such that this delete is not
+    // necessary: https://github.com/yargs/yargs/issues/1482
+    try {
+      delete argv['--']
+    } catch (_err) {}
+
+    return argv
+  }
+
+  self._runValidation = function runValidation (argv, aliases, positionalMap, parseErrors, isDefaultCommand = false) {
+    if (parseErrors) throw new YError(parseErrors.message)
+    validation.nonOptionCount(argv)
+    validation.requiredArguments(argv)
+    let failedStrictCommands = false
+    if (strictCommands) {
+      failedStrictCommands = validation.unknownCommands(argv)
+    }
+    if (strict && !failedStrictCommands) {
+      validation.unknownArguments(argv, aliases, positionalMap, isDefaultCommand)
+    }
+    validation.customChecks(argv, aliases)
+    validation.limitedChoices(argv)
+    validation.implications(argv)
+    validation.conflicting(argv)
+  }
+
+  function guessLocale () {
+    if (!detectLocale) return
+    const locale = process.env.LC_ALL || process.env.LC_MESSAGES || process.env.LANG || process.env.LANGUAGE || 'en_US'
+    self.locale(locale.replace(/[.:].*/, ''))
+  }
+
+  // an app should almost always have --version and --help,
+  // if you *really* want to disable this use .help(false)/.version(false).
+  self.help()
+  self.version()
+
+  return self
+}
+
+// rebase an absolute path to a relative one with respect to a base directory
+// exported for tests
+export function rebase (base: string, dir: string) {
+  return path.relative(base, dir)
+}
+
+/** Instance of the yargs module. */
+export interface YargsInstance {
```

Comment:
seeing all yargs' methods defined in one place is really valuable 👏

Reply:
Yes. The next step would be to add jsdoc comments here, to have overlay documentation on IDE. And the step after again would be to generate the API doc (or most of it) from these comments.